### PR TITLE
 Ticket 11, POST - /api/articles/:article_id/comments

### PR DESCRIPTION
**Pull request by KateChern**
#### In app.test.js:
The tests suite is testing  POST - /api/articles/:article_id/comments for 3 possible outcomes of the request:
- status:201, responds with the new comment object
- status:400, responds with a message 'invalid username or article' for invalid username,
- status:400, responds with a message 'Bad request' for invalid article_id,
- status:400, responds with a message 'invalid username or article' when article_id is valid but doesn't exist

#### In App.js:
 - make app.post(/api/articles/:article_id/comment) request
 - add error handler for invalid username or article
 

#### In controllers add addCommentToArticle:
- invokes `addComment` model function with `article_id` param and new comment to post
- handles errors
- sends the response to the client

#### In models add addComment :
- it sends a query to the db
- returns the posted article object

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -233,4 +233,66 @@ describe("app", () => {
...
+      body: "test",
+    };
+    return request(app)
+      .post("/api/articles/20/comments")
+      .send(testComment)
+      .expect(400)
+      .then(({ body: { msg } }) => {
+        expect(msg).toBe("invalid username or article");
+      });
+  });
```
Comment (Angaca) [inline __tests__/app.test.js:297]: Very nice testing!

File: app.js [hunk 2]
```diff
@@ -26,7 +30,11 @@ app.use((err, req, res, next) => {
...
 });
-
+app.use((err, req, res, next) => {
+  if (err.code === "23503")
+    res.status(400).send({ msg: "invalid username or article" });
+  else next(err);
+});
 app.use((err, req, res, next) => {
   console.log(err);
   res.status(500).send({ msg: "Uh oh! Server Error!" });
```
Comment (Angaca) [inline app.js:40]: Consider to move your error handling functions on a separate file

File: controllers/comments.js [hunk 3]
```diff
@@ -15,3 +15,16 @@ exports.getCommentsByArticleIdController = (req, res, next) => {
       next(err);
     });
 };
+
+exports.addCommentToArticle = (req, res, next) => {
+  const { article_id: articleId } = req.params;
+
+  addComment(req.body, articleId)
+    .then((comment) => {
+      //   console.log(comment, "controllers");
```
Comment (Angaca) [inline controllers/comments.js:24]: Remember to remove console.log or commented code from your production code

File: app.js (hunk 2 above)
Comment (KateChern) [reply app.js:40]: Can I do it on the next ticket?

File: models/comments.js [hunk 4]
```diff
@@ -8,3 +8,13 @@ exports.getCommentsByArticleId = async (articleId) => {
...
 };
+
+exports.addComment = async (newComment, articleId) => {
+  const { username, body } = newComment;
+
+  const result = await db.query(
+    `INSERT INTO comments (body, author, article_id) VALUES ($2, $1, $3) RETURNING *;`,
+    [username, body, articleId]
+  );
+  return result.rows[0];
```
Comment (Angaca) [inline models/comments.js:19]: Nice use of Async Await

**Review by Angaca — APPROVED** [review]
Very nice and clean code. Consider my comment but your code is ready to go. Amazing work with your comments in the pull request

File: app.js (hunk 2 above)
Comment (Angaca) [reply app.js:40]: 100%

**Pull request merged**